### DISCUSSION TITLE
Arbitrary length event parameters

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#2069] [Q&A]
I have an external software system which I have plumbed into an F' component which currently, as standard, sends telemetry using a single function:
```c
void
Tm_send(T_service_id  const        service_id,
        T_severity    const        severity,
        uint8_t       const *const p_data,
        uint16_t      const        data_length_bytes,
        T_destination const        destination);
```

I operate this mechanism under F' by reimplementing this function for F', rather than the original software, and having a second function (called at F' startup) which registers an instance of my F' component as the contractor for fulfilling `Tm_send()`.  `Tm_send()` then converts my severity to an F' severity and calls a component method corresponding to the severity level.

When I originally set up this system, most of the telemetry calls in the external system used a function like `Tm_send()` which, instead of accepting a buffer and data length, simply accepted a U32, but this became impractical.

Is there now a procedure for passing an arbitrarily long buffer to an F' event, along with a data length?

**Comment by LeStarch** [comment]
Is the data length bounded by some reasonable maximum? Events and Telemetry in F´ use Fw::Com buffer, which is stack allocated and applies system-wide.

Regardless, here is how you might do this in FPP.

```
array Bytes: [some maximum value] U8

event MyEvent (
    data: Bytes,
    used_size: U16
) ...
```

This creates an array type "Bytes" that holds the payload, and a used_size which carries the amount of that array that was filled.  In c++, you'd just fill each of these types before passing them to the call.

You could do something similar with string events...but those are probably null terminated and thus would not work for arbitrary data.

**Reply by JackNWhite** [reply]
Most of the telemetry is just a few bytes long, but some are over 100 bytes. I suppose I could set up two events, and choose which to use by size.
